frontend/src/App.res

```diff
@@ -69,43 +61,86 @@ let getTestMetrics = (item: GetBenchmarks.t_benchmarks): BenchmarkTest.testMetri
   }
 }
 
-let getLatestMasterIndex = (~testName, benchmarks) => {
-  BeltHelpers.Array.findIndexRev(benchmarks, (item: GetBenchmarks.t_benchmarks) => {
-    item.pull_number == None && item.test_name == testName
-  })
-}
+module BenchmarkView = {
+  let decodeRunAt = runAt => runAt->Js.Json.decodeString->Belt.Option.map(Js.Date.fromString)
 
-module BenchmarkResults = {
-  @react.component
-  let make = (~benchmarks: array<GetBenchmarks.t_benchmarks>, ~synchronize, ~repo_id) => {
-    let data = benchmarks->Belt.Array.map(getTestMetrics)
-    let selectionByTestName =
-      data->Belt.Array.reduceWithIndex(Belt.Map.String.empty, BenchmarkTest.groupByTestName)
+  let decodeMetrics = metrics =>
+    metrics
+    ->Belt.Option.getExn
+    ->Js.Json.decodeObject
+    ->Belt.Option.getExn
+    ->jsDictToMap
+    ->Belt.Map.String.map(v => BenchmarkTest.decodeMetricValue(v))
 
-    let comparisonMetricsByTestName = {
-      Belt.Map.String.mapWithKey(selectionByTestName, (testName, _) => {
-        // TODO: Use the index load the data from master and add an annotation.
-        switch getLatestMasterIndex(~testName, benchmarks) {
-        | Some(idx) => Some(benchmarks[idx]->getTestMetrics)
-        | None => None
-        }
-      })
-    }
+  let isMainBranch = pullNumber => {
+    Belt.Option.isNone(pullNumber)
+  }
 
-    let graphs = {
-      selectionByTestName
-      ->Belt.Map.String.mapWithKey((testName, testSelection) => {
-        let comparisonMetrics = Belt.Map.String.getExn(comparisonMetricsByTestName, testName)
-        <BenchmarkTest
-          ?comparisonMetrics synchronize key={testName} data testName testSelection repo_id
-        />
+  let makeBenchmarkData = (benchmarks: array<GetBenchmarks.t_benchmarks>) => {
+    benchmarks->Belt.Array.reduce(BenchmarkData.empty, (acc, item) => {
+      item.metrics
+      ->decodeMetrics
+      ->Belt.Map.String.reduce(acc, (acc, metricName, value) => {
+        BenchmarkData.add(
+          acc,
+          ~testName=item.test_name,
+          ~metricName,
+          ~runAt=item.run_at->decodeRunAt->Belt.Option.getExn,
+          ~commit=item.commit,
+          ~value,
+        )
       })
-      ->Belt.Map.String.valuesToArray
+    })
+  }
+
+  // Urql uses records to represent query results.
+  // Records have nominal types which makes two records with the same fields incompatible.
+  // This function assumes that the main benchmarks and the comparison benchmarks have the same structural type.
+  let unsafeCastComparisonBenchmarks = (
+    comparison: array<GetBenchmarks.t_comparisonBenchmarks>,
+  ): array<GetBenchmarks.t_benchmarks> => Belt.Array.map(comparison, Obj.magic)
```

Comment:
You can avoid this unsafe conversion by creating a fragment that you reuse in both places (main and comparison benchmarks).

```diff
diff --git a/frontend/src/App.res b/frontend/src/App.res
index 56189ed..2e3fa61 100644
--- a/frontend/src/App.res
+++ b/frontend/src/App.res
@@ -12,20 +12,23 @@ query {
 }
 `)
 
-module GetBenchmarks = %graphql(`
-query ($repoId: String!, $pullNumber: Int, $isMaster: Boolean!, $startDate: timestamp!, $endDate: timestamp!, $comparisonLimit: Int!) {
-  benchmarks: benchmarks(where: {_and: [{pull_number: {_eq: $pullNumber}}, {pull_number: {_is_null: $isMaster}}, {repo_id: {_eq: $repoId}}, {run_at: {_gte: $startDate}}, {run_at: {_lt: $endDate}}]}) {
+module Benchmarks = %graphql(`
+  fragment Benchmarks on benchmarks {
     run_at
     commit
     test_name
     metrics
   }
+`)
+
+module GetBenchmarks = %graphql(`
+query ($repoId: String!, $pullNumber: Int, $isMaster: Boolean!, $startDate: timestamp!, $endDate: timestamp!, $comparisonLimit: Int!) {
+  benchmarks: benchmarks(where: {_and: [{pull_number: {_eq: $pullNumber}}, {pull_number: {_is_null: $isMaster}}, {repo_id: {_eq: $repoId}}, {run_at: {_gte: $startDate}}, {run_at: {_lt: $endDate}}]}) {
+    ...Benchmarks
+  }
 
   comparisonBenchmarks: benchmarks(where: {_and: [{pull_number: {_is_null: true}}, {repo_id: {_eq: $repoId}}, {run_at: {_gte: $startDate}}, {run_at: {_lt: $endDate}}]}, limit: $comparisonLimit) {
-    run_at
-    commit
-    test_name
-    metrics
+    ...Benchmarks
   }
 }
 `)
@@ -48,7 +51,7 @@ let makeGetBenchmarksVariables = (
   }
 }
 
-let getTestMetrics = (item: GetBenchmarks.t_benchmarks): BenchmarkTest.testMetrics => {
+let getTestMetrics = (item: Benchmarks.t): BenchmarkTest.testMetrics => {
   {
     BenchmarkTest.name: item.test_name,
     metrics: item.metrics
@@ -76,7 +79,7 @@ module BenchmarkView = {
     Belt.Option.isNone(pullNumber)
   }
 
-  let makeBenchmarkData = (benchmarks: array<GetBenchmarks.t_benchmarks>) => {
+  let makeBenchmarkData = (benchmarks: array<Benchmarks.t>) => {
     benchmarks->Belt.Array.reduce(BenchmarkData.empty, (acc, item) => {
       item.metrics
       ->decodeMetrics
@@ -93,13 +96,6 @@ module BenchmarkView = {
     })
   }
 
-  // Urql uses records to represent query results.
-  // Records have nominal types which makes two records with the same fields incompatible.
-  // This function assumes that the main benchmarks and the comparison benchmarks have the same structural type.
-  let unsafeCastComparisonBenchmarks = (
-    comparison: array<GetBenchmarks.t_comparisonBenchmarks>,
-  ): array<GetBenchmarks.t_benchmarks> => Belt.Array.map(comparison, Obj.magic)
-
   @react.component
   let make = (~repoId, ~pullNumber=?, ~startDate, ~endDate) => {
     let ({ReasonUrql.Hooks.response: response}, _) = {
@@ -121,8 +117,7 @@ module BenchmarkView = {
     | PartialData(data, _) =>
       Js.log(data)
       let benchmarkDataByTestName = data.benchmarks->makeBenchmarkData
-      let comparisonBenchmarkDataByTestName =
-        data.comparisonBenchmarks->unsafeCastComparisonBenchmarks->makeBenchmarkData
+      let comparisonBenchmarkDataByTestName = data.comparisonBenchmarks->makeBenchmarkData
 
       let graphs = {
         benchmarkDataByTestName

```

Reply:
Aha! I looked for alternative approaches in Urql bindings, but didn't know about fragments. Very cool!